sources/osg/Geometry.js

```diff
@@ -59,24 +72,197 @@ Geometry.prototype = MACROUTILS.objectLibraryClass( MACROUTILS.objectInherit( No
         return this.primitives;
     },
 
-    drawImplementation: function ( state ) {
+    drawCallVAO: function ( state, vao, listVA ) {
+
+        var j, m;
+
+        var gl = state.getGraphicContext();
+
+        if ( state._currentVAO !== vao ) {
+
+            this._extVAO.bindVertexArrayOES( vao );
+
+            for ( j = 0, m = listVA.length; j < m; j++ ) {
+                if ( listVA[ j ].isDirty() )
+                    listVA[ j ].compile();
+            }
+        }
+
+        var primitives = this.primitives;
+
+        var p;
+
+        // now primitives where just IB differs
+        // TODO: find what's the usage of that thing of multiple IB?
+        // cannot think of anything that makes sense
+        // strip + primitives restart ? why no degenerate tri ?
+        for ( j = 0, m = primitives.length; j < m; ++j ) {
+
+            p = primitives[ j ];
+            if ( p.getCount() === 0 ) continue;
+            if ( p.getIndices ) {
+
+                var ib = p.getIndices();
+
+                // problem with cache on skfb (not osgjs)
+                //state.setIndexArray( ib );
+                //if ( ib !== state.currentIndexVBO ) {
+                ib.bind( gl );
+                if ( ib.isDirty() ) {
+                    ib.compile( gl );
+                }
+                state.currentIndexVBO = ib;
+                //}
+
+                p.drawElements( state, gl );
+            } else {
+                p.draw( state, gl );
+            }
+        }
+
+        //this._extVAO.bindVertexArrayOES( null );
+        state._currentVAO = vao;
+    },
+
+    drawImplementationVAO: function ( state ) {
+
```

Comment:
Current DrawVAO is ok, the version proposed doesn't add anything, and in case of "update" makes VAO recreation +  All VertexBuffer reupload at each frame (instead of just uploading modified buffers)